### PR TITLE
Add Docker config for production

### PR DESCRIPTION
**Pull request by amantri**
#60

### CODE REVIEW

File: docker/production/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ENV APP_HOME=/opt/node/app
+
+USER node:node
+WORKDIR $APP_HOME
+
+COPY --chown=node:node package*.json .
+
+# We install Husky via a "prepare" lifecycle script - disable it in prod
+# See: https://typicode.github.io/husky/guide.html#disable-husky-in-ci-docker-prod
+RUN npm pkg delete scripts.prepare
```
[the remark this answers is not in the document]
Comment (amantri) [reply docker/production/Dockerfile:15]: `--ignore-scripts` also ignores other lifecycle scripts like prepublish etc. We don't use them now, but I didn't want this to be a surprise later in case we do use it.

I think layers are inexpensive in Docker. If there is no change to package.json, then even this layer will be cached on subsequent runs.

File: docker/production/docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+services:
+  server:
+    build:
+      context: ./../..
+      dockerfile: ./docker/production/Dockerfile
+    image: unsynced/compass:latest # TODO(amantri): Set this to a shared account
+    platform: linux/amd64
+    command: bash -l -c "npm run start"
+    ports:
+      - "3000:3000"
```
[the remark this answers is not in the document]
Comment (amantri) [reply docker/production/docker-compose.yml:12]: We discussed it, and I suggested keeping it then, so as to be similar to other docker configs in our project. `docker compose build && docker compose up` works for this, assuming you have a .env.production file in the project root dir (it can be a copy of .env.local) that specifies the needed environment flags. Added a comment for environment.

[the remark this answers is not in the document]
Comment (codetheweb) [reply docker/production/docker-compose.yml:21]: sorry, I might be confused

when would you want to do `docker compose up` for this stack (in its current state)?

[the remark this answers is not in the document]
Comment (amantri) [reply docker/production/docker-compose.yml:21]: To test it locally. I used that to debug issues on the image. TBF, `docker compase up` is nice to have.

`docker compose build` is easier to work with than `docker build` - I tried (not very hard) to get `docker build` to work, but minor things like context etc. were not being correctly set, so I gave up on that and stuck with `docker compose`. I can try harder to get `docker build` to work :)

[the remark this answers is not in the document]
Comment (codetheweb) [reply docker/production/docker-compose.yml:21]: I agree it's useful to have for debugging images, but in that case wouldn't you want a way of bringing up the rest of the stack (Postgres) as well?

forgot to mention it earlier but it's convention to have `Dockerfile` at the root unless impossible which would fix the context issue when building

also not sure if you saw yet but I removed the docker dev workflow in my PR: https://github.com/sfbrigade/compass/pull/85

[the remark this answers is not in the document]
Comment (amantri) [reply docker/production/docker-compose.yml:21]: Ah ok - just saw the changes on #85. Removing option 1 is fine, but we do need to keep the docker configs for prod and dev, so it might make sense to keep the docker directory.

Re Postgres and other services in the prod image: specifying DATABASE_URL and related vars in .env.production will allow you to run the image (i.e. option 3). I didn't include postgres in the prod version as it is not really needed for the actual production environment.

[the remark this answers is not in the document]
Comment (codetheweb) [reply docker/production/docker-compose.yml:21]: > I didn't include postgres in the prod version as it is not really needed for the actual production environment.

Maybe we're talking about different things, I was advocating for having a stack that matches production services, not one that connects to them.

> Removing option 1 is fine, but we do need to keep the docker configs for prod and dev, so it might make sense to keep the docker directory.

At that point we'll only have 1 Dockerfile, so I would push for it being at the repo root to match convention. I can do that in my PR instead if this gets merged first since there's still 2 Dockerfiles in this PR context.

[the remark this answers is not in the document]
Comment (amantri) [reply docker/production/docker-compose.yml:21]: > Maybe we're talking about different things, I was advocating for having a stack that matches production services, not one that connects to them.

I think we are talking about the same thing. Just to be sure, I am not suggesting that an engineer connect to prod services from a local server - in fact, the networking settings won't allow that. To run the production image, you would need to bring up postgres and other services locally (old option 2 or 3) and connect to them via env vars. This is basically the same as the dev stack where we run the Compass server locally, and connect to postgres, but instead of running the dev binary, you will be running the prod binary. We can share the docker-compose.yml for the supporting files.


> At that point we'll only have 1 Dockerfile, so I would push for it being at the repo root to match convention. I can do that in my PR instead if this gets merged first since there's still 2 Dockerfiles in this PR context.

Yes, but it might still be better to differentiate them via some folder structure as the Dockerfile will be for production Compass binary, while the docker-compose.yml will be for the supporting services. Putting them in the same directory suggests that they are related, while in reality they aren't. Maybe the current "docker/postgres" folder should be renamed to "docker/3p_services" or similar to indicate the supporting services.

[the remark this answers is not in the document]
Comment (codetheweb) [reply docker/production/docker-compose.yml:21]: > This is basically the same as the dev stack where we run the Compass server locally, and connect to postgres, but instead of running the dev binary, you will be running the prod binary. We can share the docker-compose.yml for the supporting services.

What's the difference between the dev and prod service version in this context? Wouldn't we just have one prod Dockerfile?

[the remark this answers is not in the document]
Comment (amantri) [reply docker/production/docker-compose.yml:21]: Debug build vs optimized prod build of our app (Compass)

[the remark this answers is not in the document]
Comment (amantri) [reply docker/production/docker-compose.yml]: Updated the PR per our discussion yesterday. Changes:
 - Running db:migration in CMD (now using shell form instead of exec form as there are multiple commands)
 - moved docker-compose.yml to supporting_services/

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+#
+# To run the production image locally, create a .env.production file with the
+# appropriate env vars (a copy of .env.local works - see NOTE below).
+#
+# Command to build and run a Docker image:
+#  % docker build . -t compass && docker run -p=3000:3000 --name=compass --rm --env-file=.env.production compass
+#
+# NOTE: If you have the references to `localhost` in the .env file for other
+# services like postgres, replace `localhost` with the local IP address, e.g.
+# using `% ipconfig getifaddr en0`. This is needed as the docker image run in
```
Comment (codetheweb) [inline Dockerfile:11]: `host.docker.internal` also resolves to the host IP

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+# its own VM and `localhost` resolves to to is own VM, not the host.
+
+# Start with the latest Node.js LTS release
+FROM --platform=linux/amd64 node:18-bullseye-slim
```
Comment (codetheweb) [inline Dockerfile:15]: does the arch need to be specified? makes it slower for people using arm machines

File: src/backend/db/lib/migrate.ts [hunk 5]
```diff
@@ -23,16 +23,20 @@ export const migrate = async (
...
+    host: connectionConfig.host ?? "",
+    port: connectionConfig.port ? parseInt(connectionConfig.port, 10) : 5432,
+
+    database: connectionConfig.database ?? "",
+    ensureDatabaseExists: true,
+    defaultDatabase: "postgres",
+  };
+
+  // Run migrations. Creates database if needed.
+  await postgresMigrations.migrate(dbConfig, migrationsDirectory);
```
Comment (codetheweb) [inline src/backend/db/lib/migrate.ts:39]: why were these changes needed?

File: src/backend/db/lib/reset.ts [hunk 6]
```diff
@@ -32,9 +32,6 @@ export const reset = async (databaseUrl: string) => {
 
       throw error;
     });
-
-  logger.info(`Creating database ${targetDatabaseName}...`);
-  await client.query(`CREATE DATABASE ${targetDatabaseName}`);
```
[the remark this answers is not in the document]
Comment (amantri) [reply src/backend/db/lib/reset.ts:37]: Yes. The create db is now handled in migrate.ts

[the remark this answers is not in the document]
Comment (codetheweb) [reply src/backend/db/lib/reset.ts:37]: 👍 didn't realize `ensureDatabaseExists` is an available parameter

File: Dockerfile (hunk 4 above)
Comment (amantri) [reply Dockerfile:15]: Unfortunately yes - AWS needs amd64 images. 

Do you envision using the prod images locally on a regular basis or only for debugging the image? If the latter, the slowness might be tolerable. If the former, we can look into creating multi-platform images.

File: Dockerfile (hunk 3 above)
Comment (amantri) [reply Dockerfile:11]: Updated documentation.

File: Dockerfile (hunk 4 above)
Comment (amantri) [reply Dockerfile:16]: Another option to build the image for arm is to specify the --platform in the CLI, e.g. `docker build --platform linux/arm64 .`

Comment (codetheweb) [reply Dockerfile:16]: if this is omitted, it will default to whatever the host machine is--I would recommend not specifying it here and passing it as `docker build --platform=linux/amd64` if building for x86 on ARM.

**Review by codetheweb — APPROVED** [review]

File: Dockerfile (hunk 4 above)
Comment (codetheweb) [reply Dockerfile:20]: AWS has ARM instances available, often for cheaper than x86; so defaulting to x86 doesn't really make sense to me. But 🤷

Comment (francisli) [reply Dockerfile:20]: FYI if you remove the platform flag you can use `docker buildx` to create and push multi architecture images to ECR in one step that contains both x86 and ARM layers, so that you can use either platform on AWS.
